### PR TITLE
Use CircleCI

### LINKED ISSUE #1692: Consider different CI provider
Our current CI is completely dependent on github actions. Github actions only allows for [5 concurrent MacOS jobs ](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits)**per organization**. This limit cannot be increased unless we migrate to Github Enterprise.

Currently we are competing for workers with both Scarb and other software mansion projects unrelated to starknet.

We should consider an alternative CI provider with better (more configurable) parallel job execution and migrate some/all of our actions to it.

### LINKED ISSUE #1745: Optimize CI

### LINKED ISSUE #1778: Scripts for test releases
I would like to have a script which creates a release which can be downloaded with protostar curl but without releasing docs and without bumping the version number. It would be useful for debugging

### PR DESCRIPTION
**Pull request by THenry14**
Before/right after the merge:
[X] Configure repo to allow circle (software-mansion org admin only)
[ ] Configure project on circle with all the necessary project variables (tokens, slack_url etc)
[ ] Change branch protection rules to await circle checks

Closes: https://github.com/software-mansion/protostar/issues/1692
Closes: https://github.com/software-mansion/protostar/issues/1745
Closes: https://github.com/software-mansion/protostar/issues/1778

### Checklist
- [X] Relevant issue is linked
- [X] Docs updated/issue for docs created
- [X] Added relevant tests

### CODE REVIEW

**Comment by THenry14** [comment]
> In #1881 we add new bindigns + script which we will need to include in CI @karol-bisztyga @THenry14 can you sync on that?

Not a problem, should I rebase on it right now, or is it going to be merged _soon_ ?

**Comment by karol-bisztyga** [comment]
> > In #1881 we add new bindigns + script which we will need to include in CI @karol-bisztyga @THenry14 can you sync on that?
> 
> Not a problem, should I rebase on it right now, or is it going to be merged _soon_ ?

Maybe today but I don't want to keep you blocked, so rebasing is probably a good idea.

File: scripts/install_cairo_bindings.sh [hunk 1]
```diff
@@ -5,37 +5,33 @@ set -e
...
+  if [[ $(uname -s) == 'Darwin' ]]; then
     CFLAGS=-I/opt/homebrew/opt/gmp/include LDFLAGS=-L/opt/homebrew/opt/gmp/lib poetry install
   else
     poetry install
   fi
 fi
 
 function install_dev() {
-  git pull --recurse-submodules
+  git pull origin $(git rev-parse --abbrev-ref HEAD) --ff-only --recurse-submodules
```
Comment (karol-bisztyga) [inline scripts/install_cairo_bindings.sh:16]: Could you please explain why we need to do it like this?

File: scripts/install_cairo_bindings.sh [hunk 2]
```diff
@@ -5,37 +5,33 @@ set -e
...
 
-  pushd cairo
-  pushd crates/cairo-lang-python-bindings
+  pushd cairo/crates/cairo-lang-python-bindings
   rustup override set nightly-2022-11-03 || return 1;
   maturin build || return 1;
-  popd # crates/cairo-lang-python-bindings
+  popd
```
Comment (karol-bisztyga) [inline scripts/install_cairo_bindings.sh:30]:
```suggestion
  popd # cairo/crates/cairo-lang-python-bindings
```
makes it easier to read

File: scripts/install_cairo_bindings.sh [hunk 3]
```diff
@@ -5,37 +5,33 @@ set -e
...
 
-  pushd target/wheels
+  pushd cairo/target/wheels
   pip install "./$(ls | grep cairo_python_bindings)" || return 1;
-  popd # target/wheels
-  popd # cairo
+  popd
```
Comment (karol-bisztyga) [inline scripts/install_cairo_bindings.sh:34]:
```suggestion
  popd # cairo/target/wheels
```

File: scripts/install_cairo_bindings.sh (hunk 1 above)
Comment (THenry14) [reply scripts/install_cairo_bindings.sh:16]: We need to be able to build from users branches, in order to provide test and pre releases. 
The `git rev-parse --abbrev-ref HEAD` gets the current branch name.

**Review by MaksymilianDemitraszek — APPROVED** [review]

**Review by bartekryba — APPROVED** [review]
Lgtm, didn't find anything that would catch my eye.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+      platform:
+        type: enum
+        enum: ["linux", "mac"]
+    steps:
+      - run:
+          name: Install packages
+          command: |
+            poetry config installer.modern-installation false
+            if [[ << parameters.platform >> == "mac" ]]; then
+              CFLAGS=-I/opt/homebrew/opt/gmp/include LDFLAGS=-L/opt/homebrew/opt/gmp/lib poetry --no-ansi install
```
[the remark this answers is not in the document]
Comment (THenry14) [reply .circleci/config.yml]: good idea, thanks! added in newest commit

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+          paths:
+            - ~/.cache/yarn
+      - run:
+          name: Pack docs
+          command: tar -czvf protostar-docs.tar.gz website/build
+      - persist_to_workspace:
+          root: .
+          paths: protostar-docs.tar.gz
+
+  # TODO: it seems gh-pages can have problems with already deployed gh-pages branches? to investigate + change email
```
[the remark this answers is not in the document]
Comment (THenry14) [reply .circleci/config.yml]: Sure, I will remember to do it this way from now on.
Here's the task https://github.com/software-mansion/protostar/issues/1926

**Pull request merged**